### ISSUE TITLE
Slow consumer flow

### ISSUE THREAD

**Issue by ctrochalakis** [confluentinc/confluent-kafka-go#55]
Hello Magnus,

I am building a kafka consumer using the Go bindings, I have some questions regarding
multiplexing control and data events in a single channel (`Events()`).

Setup:

  * Slow consumer, one job at a time (a job can take up to 10min to process)
  * I am using the channel based consumer API
  * Manual commits per message

Our processing function is directly hooked to a consumer.Events() loop along with handlers
for assigned & revoked partitions. So, since the call can block for up to 10min the `Events()`
channel buffer might fill up. Is this something to worry about?

Also, what is going to happen if e.g. a partition is revoked or assigned but that event is stuck behind a set of `kafka.Message`s? Those control messages are going to be delivered after several minutes.

So, basically my question is what's the correct way to implement a slow-consumer.

Thank you,
Chris

**Comment by edenhill** [comment]
You are right that the current design of combining messages and events on the same channel or queue might lead to problems with slow consumers.

With librdkafka it is possible redirect messages to its own queue and thus separate messages and events, avoiding messages to (head-of-line-)block sub-sequent events.

Let me look into what it would like to provide the same functionality in Go, it would probably boil down to something like:
`Assign(partitions, channel_for_msgs_only)`

> Also, what is going to happen if e.g. a partition is revoked or assigned but that event is stuck behind a set of kafka.Messages? Those control messages are going to be delivered after several minutes.

Any queued messages will be purged when a rebalance takes place, so if you are using the Poll() interface the next call to poll() should give you an assign/revoke event.
If you are using the channel consumer then the Events channel might have some messages queued (depending on channel size, but typically at least one message).

**Comment by JeanMertz** [comment]
Posting my support for this feature here as well. For another situation where this is applicable, see https://github.com/confluentinc/confluent-kafka-go/issues/65#issuecomment-368433290.

**Comment by milindl** [comment]
Given that the channel-based consumer has been deprecated for a while, I'm closing the ticket, especially as the issue of data messages before control messages isn't there anymore since Poll() purges queued messages. The issues related to delayed control messages, or the channel buffer filling up, won't be there anymore.

As long as max.poll.interval.ms is configured correctly, and Poll or ReadMessage is called at least as frequently as that, it should be OK to take as much time as needed to process the message.
